### PR TITLE
crd: add a flag to specify the scope of the crd

### LINKED ISSUE #890: Scaffolding cluster scoped CRDs
## Feature Request
operator-sdk currently (v0.3.0) supports scaffolding namespaced CRDs only. Please support scaffolding cluster scoped CRDs too. Btw kubebuilder supports it and rationale is somewhat documented on http://kubebuilder.netlify.com/basics/what_is_a_resource.html#namespaces

For consistency with already supported scaffolding of cluster wide watching operators (via `--cluster-scoped` switch in `operator-sdk new` command), scaffolding of cluster scoped CRDs should probably also be supported with similar `--cluster-scoped` switch for `operator-sdk add api` command.

### PR DESCRIPTION
**Pull request by xcoulon**
**Description of the change:**
- add a `--cluster-scoped-crd` flag in the `add api`
and `add crd` commands
- rename the `--cluster-scoped` flag to `-cluster-scoped-operator`
in the `new` command, and add a `--cluster-scoped-crd` flag, but
only for the `ansible` and `helm` types or operators

**Motivation for the change:**
See https://github.com/operator-framework/operator-sdk/issues/890#issuecomment-452809400

Fixes #890

Signed-off-by: Xavier Coulon <xcoulon@redhat.com>

### CODE REVIEW

File: commands/operator-sdk/cmd/new.go [hunk 1]
```diff
@@ -288,9 +292,9 @@ func doHelmScaffold() error {
 			ChartName: chart.GetMetadata().GetName(),
 		},
 		&scaffold.ServiceAccount{},
-		&scaffold.Role{IsClusterScoped: isClusterScoped},
-		&scaffold.RoleBinding{IsClusterScoped: isClusterScoped},
-		&helm.Operator{IsClusterScoped: isClusterScoped},
+		&scaffold.Role{IsClusterScoped: isClusterScopedOperator},
+		&scaffold.RoleBinding{IsClusterScoped: isClusterScopedOperator},
+		&helm.Operator{IsClusterScoped: isClusterScopedOperator},
 		&scaffold.CRD{Resource: resource},
```
Comment (estroz) [inline commands/operator-sdk/cmd/new.go]: This also should receive the `isClusterScoped` variable correct?

File: pkg/scaffold/crd.go [hunk 2]
```diff
@@ -145,25 +151,30 @@ func (s *CRD) CustomRender() ([]byte, error) {
 	return k8sutil.GetObjectBytes(dstCRD)
 }
 
-func newCRDForResource(r *Resource) *apiextv1beta1.CustomResourceDefinition {
-	return &apiextv1beta1.CustomResourceDefinition{
+func (s *CRD) newCRD() (*apiextv1beta1.CustomResourceDefinition, error) {
```
Comment (estroz) [inline pkg/scaffold/crd.go]: No need to return an error.

File: pkg/scaffold/operator.go [hunk 3]
```diff
@@ -24,7 +24,6 @@ const OperatorYamlFile = "operator.yaml"
 
 type Operator struct {
 	input.Input
-
```
Comment (estroz) [inline pkg/scaffold/operator.go]: Keep this line separator.

File: pkg/scaffold/scaffold.go [hunk 4]
```diff
@@ -181,7 +181,7 @@ func (s *Scaffold) doRender(i input.Input, e input.File, absPath string) error {
 		}
 	}
 	_, err = f.Write(b)
-	log.Infoln("Created", i.Path)
+	log.Infof("Created '%s' (%s)\n", i.Path, absPath)
```
Comment (estroz) [inline pkg/scaffold/scaffold.go]: Undo this change (outside of the scope of this PR).

File: test/e2e/doc.go [hunk 5]
```diff
@@ -0,0 +1,16 @@
+// Copyright 2019 The Operator-SDK Authors
```
Comment (estroz) [inline test/e2e/doc.go]: This file isn't necessary, as everything in `test/` is for SDK testing and not exported.

File: test/test-framework/test/e2e/doc.go [hunk 6]
```diff
@@ -0,0 +1,16 @@
+// Copyright 2019 The Operator-SDK Authors
```
Comment (estroz) [inline test/test-framework/test/e2e/doc.go]: This file isn't necessary, as everything in `test/` is for SDK testing and not exported.

File: pkg/scaffold/operator.go (hunk 3 above)
Comment (xcoulon) [reply pkg/scaffold/operator.go]: ok

File: pkg/scaffold/crd.go (hunk 2 above)
Comment (xcoulon) [reply pkg/scaffold/crd.go]: right, this is a left-over form a previous idea, and I actually don't need it anymore

File: commands/operator-sdk/cmd/new.go (hunk 1 above)
Comment (xcoulon) [reply commands/operator-sdk/cmd/new.go]: @estroz ah yes, indeed!

Comment (xcoulon) [reply commands/operator-sdk/cmd/new.go]: fixed in 543b7d0

File: pkg/scaffold/crd.go (hunk 2 above)
Comment (xcoulon) [reply pkg/scaffold/crd.go]: fixed in 543b7d0

File: pkg/scaffold/operator.go (hunk 3 above)
Comment (xcoulon) [reply pkg/scaffold/operator.go]: fixed in 543b7d0

File: pkg/scaffold/scaffold.go (hunk 4 above)
Comment (xcoulon) [reply pkg/scaffold/scaffold.go]: sure, let me revert it

File: test/e2e/doc.go (hunk 5 above)
Comment (xcoulon) [reply test/e2e/doc.go]: well, I had some errors reported by VSCode because there were no non-test go files in those packages, hence these 2 `doc.go` files, but I can remove them if you prefer.

File: test/test-framework/test/e2e/doc.go (hunk 6 above)
Comment (xcoulon) [reply test/test-framework/test/e2e/doc.go]: See comment above

File: pkg/scaffold/scaffold.go (hunk 4 above)
Comment (xcoulon) [reply pkg/scaffold/scaffold.go]: done in 1c31dac

File: test/e2e/doc.go (hunk 5 above)
Comment (xcoulon) [reply test/e2e/doc.go]: done in 2dfd90f

File: test/test-framework/test/e2e/doc.go (hunk 6 above)
Comment (xcoulon) [reply test/test-framework/test/e2e/doc.go]: done in 2dfd90f

**Comment by xcoulon** [comment]
@estroz I believe that I addressed all your comments. Please let me know if you have further concerns 

also, my apologies for the several pushes to fix the remaining compilation errors! I should have run `make install` well before! 😬

**Comment by hasbro17** [comment]
@xcoulon Thanks for the PR. I think this is pretty much in line with what we initially proposed on #890 but per my latest comment I think we still need to reach a consensus on some of the edge cases with these two flags before we can move forward with these changes.

**Comment by xcoulon** [comment]
> @xcoulon Thanks for the PR. I think this is pretty much in line with what we initially proposed on #890 but per my latest comment I think we still need to reach a consensus on some of the edge cases with these two flags before we can move forward with these changes.

Thanks for your feedback, @hasbro17! I'll be happy to add more changes in this PR once the consensus is reached, in order to cover the edge cases ;)

**Comment by LiliC** [comment]
@hasbro17 Do we still want this PR in?

**Comment by hasbro17** [comment]
Forgot to follow up on this. Thanks for reminding me @LiliC (and @openshift-ci-robot  😅 )

No, we probably don't want to configure the Operator or CRD scope from the CLI.
Per #890 and #1434 we're going to leave those changes to the user to keep the workflow simple.

@xcoulon Thanks for the PR and discussion but I'll close this for now.

**Comment by xcoulon** [comment]
> Forgot to follow up on this. Thanks for reminding me @LiliC (and @openshift-ci-robot 😅 )
> 
> No, we probably don't want to configure the Operator or CRD scope from the CLI.
> Per #890 and #1434 we're going to leave those changes to the user to keep the workflow simple.
> 
> @xcoulon Thanks for the PR and discussion but I'll close this for now.

No worries, it was my pleasure to try to help here ;)
